### PR TITLE
[WOR-1584] Merge attributes during Azure workspace clone

### PR DESCRIPTION
**Pull request by cahrens**
Ticket: https://broadworkbench.atlassian.net/browse/WOR-1584

This bug only existed on Azure, but I added test coverage for it in the GCP case as well (as I couldn't find any).

Tested manually with a locally running Rawls.

---

**PR checklist**

- [ ] Include the JIRA issue number in the PR description and title
- [ ] Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] If you changed anything in `model/`, then you should [publish a new official `rawls-model`](https://github.com/broadinstitute/rawls/blob/develop/README.md#publish-rawls-model) and update `rawls-model` in [Orchestration's dependencies](https://github.com/broadinstitute/firecloud-orchestration/blob/develop/project/Dependencies.scala).
- [ ] Get two thumbsworth of PR review
- [ ] Verify all tests go green, including CI tests
- [ ] **Squash commits and merge** to develop (branches are automatically deleted after merging)
- [ ] Inform other teams of any substantial changes via Slack and/or email

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 1]
```diff
@@ -363,10 +363,12 @@ class MultiCloudWorkspaceService(override val ctx: RawlsRequestContext,
 
     assertBillingProfileCreationDate(profile)
     val workspaceId = UUID.randomUUID()
+    // Merge together source workspace and destination request attributes
+    val mergedRequest = request.copy(attributes = sourceWorkspace.attributes ++ request.attributes)
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:367]: This is the asynchronous clone path.

To match the existing code in the GCP case: 
https://github.com/broadinstitute/rawls/blob/44f717ce2b7d076aebf577b64377c7c56e4e11f3/core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala#L1112

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala [hunk 2]
```diff
@@ -464,10 +466,12 @@ class MultiCloudWorkspaceService(override val ctx: RawlsRequestContext,
     val wsmConfig = multiCloudWorkspaceConfig.workspaceManager
     val workspaceId = UUID.randomUUID()
 
+    // Merge together source workspace and destination request attributes
+    val mergedRequest = request.copy(attributes = sourceWorkspace.attributes ++ request.attributes)
```
Comment (cahrens) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceService.scala:470]: This is the synchronous clone path.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala [hunk 3]
```diff
@@ -1651,7 +1651,9 @@ trait TestDriverComponent extends DriverComponent with DataAccess with DefaultIn
       createdDate = currentTime(),
       lastModified = currentTime(),
       createdBy = "testUser",
-      attributes = Map.empty,
+      attributes = Map(
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/TestDriverComponent.scala:1654]: Common test workspace used in the particular test that I extended.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceSpec.scala [hunk 4]
```diff
@@ -1335,7 +1335,7 @@ class MultiCloudWorkspaceServiceSpec extends AnyFlatSpec with Matchers with Opti
 
   it should
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceSpec.scala:1336]: This is a test for our synchronous clone path.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceUnitTestsSpec.scala [hunk 5]
```diff
@@ -312,4 +314,90 @@ class MultiCloudWorkspaceServiceUnitTestsSpec
     verify(workspaceManagerResourceMonitorRecordDao).create(any())
   }
 
+  it should "merge together source and destination attributes" in {
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceUnitTestsSpec.scala:317]: Async clone workspace tests live in this file and are more fully mocked.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceUnitTestsSpec.scala [hunk 6]
```diff
@@ -312,4 +314,90 @@ class MultiCloudWorkspaceServiceUnitTestsSpec
...
+    val destWorkspace = mock[Workspace]
+    val mergedAttributes = Map(
+      AttributeName.withDefaultNS("description") -> AttributeString("source description"),
+      AttributeName.withDefaultNS("destination") -> AttributeString("destination only")
+    )
+    val mergedWorkspaceRequest =
+      WorkspaceRequest("dest-namespace", "dest-name", mergedAttributes, policies = Some(policies))
+    doReturn(Future.successful(destWorkspace))
+      .when(service)
+      .createNewWorkspaceRecord(
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/MultiCloudWorkspaceServiceUnitTestsSpec.scala:389]: This is the actual point that fails without the fix… we never receive a `createNewWorkspaceRecord` call with the merged request attributes, and thus the mock returns null and the code blows up.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala [hunk 7]
```diff
@@ -2335,6 +2335,7 @@ class WorkspaceServiceSpec
     workspace.workspaceVersion should be(WorkspaceVersions.V2)
     workspace.googleProjectId.value should not be empty
     workspace.googleProjectNumber should not be empty
+    workspace.attributes should be(baseWorkspace.attributes)
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:2338]: This file has the GCP workspace clone tests. The source destination workspace (`baseWorkspace`) already had attributes defined, but there was no test coverage that they were copied into the destination workspace.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala [hunk 8]
```diff
@@ -2366,6 +2367,36 @@ class WorkspaceServiceSpec
     workspace.googleProjectNumber should not be empty
   }
 
+  it should "merge destination attributes with source attributes" in withTestDataServices { services =>
+    val baseWorkspace = testData.workspace
+    val newWorkspaceName = "cloned_space"
+    testData.workspace.attributes.get(AttributeName.withDefaultNS("string")).value should be(
+      AttributeString("yep, it's a string")
```
Comment (cahrens) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala:2374]: Verifying that the source test workspace had a value that we are then able to override via the clone request.

**Review by blakery — APPROVED** [review]

**Review by marctalbott — APPROVED** [review]

**Pull request merged**